torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -419,6 +419,41 @@ class ShapePropagator {
     setUnshapedType(cat_node);
   }
 
+  void propagateTorchTensorShape(Node* node) {
+    auto input_type = node->inputs().at(0)->type();
+
+    size_t dims = 0;
+    auto input_base_type = input_type;
+    auto list_type = input_type->cast<ListType>();
+    while (list_type) {
+      dims++;
+      input_base_type = list_type->getElementType();
+      list_type = input_base_type->cast<ListType>();
+    }
+
+    at::ScalarType default_type = scalarTypeFromJitType(input_base_type);
+    if (auto grad_index = node->schema().argumentIndexWithName("dtype")) {
+      auto inp = toIValue(node->inputs().at(*grad_index));
+      if (inp == c10::nullopt) {
+        return;
+      } else if (!inp->isNone()) {
+        default_type = inp->toScalarType();
+      }
+    }
+
+    at::Device default_device = at::kCPU;
+    if (auto device_index = node->schema().argumentIndexWithName("device")) {
+      auto inp = toIValue(node->inputs().at(*device_index));
+      if (inp == c10::nullopt) {
+        return;
+      } else if (inp->isDevice()) {
```

Comment:
shouldn't this have a `isNone()` check since its also an optional argument?

Reply:
The reason the one above this is `inp->isNone()` is because ScalarType is represented by ints, so there is no `is->ScalarType`. I will switch this check to `inp->isNone()` as well so it is more readable, but it doesn't change the behavior.